Week3/README.md

```diff
@@ -20,77 +20,93 @@ Your teachers Igor and Unmesh have made video lectures for this week's material.
 
 ## 1. Normalization and normal forms
 
-When setting up a database, one of the most important aspects to be taken into account is to ensure that duplication of data is minimized. This is done for 2 purposes:
+When setting up a database, one of the most important aspects to be taken into account is to ensure that duplication of data is minimized.
+If you do store data twice in your database, but on update forget to update all locations you will end up with a data conflict.
 
-- Reducing the amount of storage needed to store the data.
-- Avoiding unnecessary data conflicts that may creep in because of multiple copies of the same data getting stored.
+A second advantage of a fully normalized database is that it allows its structure to be extended later to accommodate new types of data without changing existing structure too much.
+As a result, code interacting with the database is minimally affected.
 
 ![if you dont do database normalization you will have a bad time](https://i.imgflip.com/1uidqc.jpg)
 
 Database normalization is a technique that helps in optimally designing the database to ensure the above points. The core idea of database normalization is to divide large tables into smaller subtables and store pointers to data rather than replicating it.
 
 There are various database “Normal” forms. Each normal form (NF) has an importance which helps in optimizing the database to save storage and to reduce redundancies. These normal forms build incrementally. e.g. a database is in 3NF if it is already in 2NF and satisfied the rules for 3rd normal form.
+The normal forms are written abstractly in a hard-to-understand formal language, we will keep it short and practical. 
 
-#### 1st normal form (1NF) (4 rules)
+#### 1st normal form (1NF)
 
-The rules for first normal form are:
+The first normal form requires your attributes to stay proper, in essence coming down to this list:
 
-- Rule 1 : Single valued attributes (each column should have atomic value, no multiple values)
-- Rule 2 : Attribute domain should not change
-- Rule 3 : Unique names for attributes / columns
-- Rule 4 : Order does not matter
-- Rule 5 : Repeating columns are not allowed (ie TelephoneNumber1, TelephoneNumber2, etc.)
-
-Did you get that? Neither did I! These normal form rules are written abstractly in a hard-to-understand formal language.
```

Comment:
I did. But main reason I wrote them formally is that students will end up reading technical stuff online anyway. And teachers should be able to explain what they mean.

Reply:
>> Did you get that? Neither did I!
>
> I did.

_Did you just answer your own question, again? 😛_

Reply:
At first I didn't feel the need to rewrite the 1NF rules but mainly wanted to rewrite 2NF and 3NF.
Those two seemed way easier to explain in common English compared to theoretical definitions.
So that's why I also rewrote 1NF to stay practical for the whole normalization topic.

I'm personally a big fan of practical understanding, that helps me getting into the essence of a subject.
And I find it also easier to let students get a feeling for a subject when I teach it more practical.
That why your argument look like this to me:
"They'll read other sources online that are confusing anyway so our explanation can be confusing too"
😉 

> And teachers should be able to explain what they mean.

I agree, but wouldn't it be great if the students would grasp it straight away?
Other online sources are also easier to understand that way because they start with a basic understanding.

Reply:
Outside of the discussion on _formal vs practical teaching_;
I removed rules 2, 3 and 4 for brevity

>Rule 2 : Attribute domain should not change

I couldn't find any proper explanation of this rule in HYF or in the given online sources.
And this one does need some explanation if you'd ask me.

> Rule 3 : Unique names for attributes / columns
> Rule 4 : Order does not matter

Why bother mentioning these two?
Since most databases will "apply" these two rules out of the box. 
These two rules only seem valuable in a theoretical case.

And I did add a new rule "Primary key (no duplicate tuples)"

Reply:
> > > Did you get that? Neither did I!
> > 
> > 
> > I did.
> 
> _Did you just answer your own question, again? 😛_

Yeah, It was sarcastic.

Reply:
> At first I didn't feel the need to rewrite the 1NF rules but mainly wanted to rewrite 2NF and 3NF.
> Those two seemed way easier to explain in common English compared to theoretical definitions.
> So that's why I also rewrote 1NF to stay practical for the whole normalization topic.
> 
> I'm personally a big fan of practical understanding, that helps me getting into the essence of a subject.
> And I find it also easier to let students get a feeling for a subject when I teach it more practical.
> That why your argument look like this to me:
> _"They'll read other sources online that are confusing anyway so our explanation can be confusing too"_
> 😉
> 
> > And teachers should be able to explain what they mean.
> 
> I agree, but wouldn't it be great if the students would grasp it straight away?
> Other online sources are also easier to understand that way because they start with a basic understanding.

My argument is "write what they will find in online resources (formal language) and then explain it in a way they can understand".

Reply:
These rules for 1NF aren't that hard so it's not a big deal to me. 
I'll revert them back to formal language.

But when searching online there was only one source (studytonight.com) that listed the rules that where listed here.
The most common one to find is "atomic values", other sources add "prevent repeating columns" and others (like wikipedia) add "prevent duplicate tuples"

Reply:
As long as we cover all correct versions of these rules, I am fine with it.

Reply:
I am with Tomas here that there isn't really a widely used formal description (just the acronyms) and that different sites use different explanations. Which means I think we are free to give our own explanation of the rules. I do think all of them should be there though, even if the database checks them.

So my vote atm would be to use the new descriptions and add the 2 removed ones back.

Reply:
After hours of googling on the topic, I hardly found any sources online that mention these rules; only on studytonight.com did. That's why I proposed to remove the three rules. (you can find sources if you search for the exact words, but you don't come across them when you search 'database normalization')

About these two: `Unique names for columns` and `Order does not matter`
I don't see the value of adding rules that can be ignored. But these two are so simple so I don't mind either to keep them in, they're easy to read and forget. So I'd coincide for sake of keeping the argument brief.

About this one: `Rule 2 : Attribute domain should not change`
This one isn't super straight forward to me and it isn't explained in our readme.
The explanation from studytonight in plain english is like this: _"This is more of a "Common Sense" rule. In each column the values stored must be of the same kind or type._
Putting that in our docs also doesn't make sense to me. The example would be "You shouldn't put dates in a varchar column".
That's common sense, even to non-IT students, right?

Reply:
Another argument would be like the one about the exercises: https://github.com/HackYourFuture/databases/pull/460

>I think we should not focus on the exact normalization steps in the exercise.
The goal however seems legit, so I propose to keep the exercise but stress less about the normal forms.

In this case I would make the same point, to teach them "how to basically normalize" in stead of "learn all the rules"

Reply:
> After hours of googling on the topic, I hardly found any sources online that mention these rules; only on studytonight.com did. That's why I proposed to remove the three rules. (you can find sources if you search for the exact words, but you don't come across them when you search 'database normalization')
> 
> About these two: `Unique names for columns` and `Order does not matter`
> I don't see the value of adding rules that can be ignored. But these two are so simple so I don't mind either to keep them in, they're easy to read and forget. So I'd coincide for sake of keeping the argument brief.
> 
> About this one: `Rule 2 : Attribute domain should not change`
> This one isn't super straight forward to me and it isn't explained in our readme.
> The explanation from studytonight in plain english is like this: _"This is more of a "Common Sense" rule. In each column the values stored must be of the same kind or type._
> Putting that in our docs also doesn't make sense to me. The example would be "You shouldn't put dates in a varchar column".
> That's common sense, even to non-IT students, right?

Common sense is not so common ;-) 
One can easily insert date in a varchar column because both support string literals. When prompting command (say `insert into`) in the class, I remember having to spell-out `INSERT`  SPACE  `INTO` 

Then again, if these rules are not that useful, we can remove them.

Reply:
> Another argument would be like the one about the exercises: #460
> 
> > I think we should not focus on the exact normalization steps in the exercise.
> > The goal however seems legit, so I propose to keep the exercise but stress less about the normal forms.
> 
> In this case I would make the same point, to teach them "how to basically normalize" in stead of "learn all the rules"

I see rules as `tools to explain why some decisions were made`. If students know and understand what is a `transitive dependency` or a `partial dependency`, then they can argue about how that is violated instead of handwaving arguments (this would come handy in interviews when knowing technical terms/formal language is a plus IMHO)

In summary, I think we can simplify things not by removing rules but by adding the extra explanations where the rule are too formal/too hard to understand.

Reply:
I didn't think of the value that it can have in interviews and debates. Perhaps because I've never had such interviews and when I interview someone I wouldn't ask such questions. 😆

That said, I have restored the rule about "consistent domain values" and also added it to the exercise
https://github.com/HackYourFuture/databases/pull/458/commits/b648aaf6082992d6cde4af0f6fa6230cd7b3869a